File: app_manager/appmanager/src/main/java/app/editors/manager/ui/fragments/main/DocsRecentFragment.java
package app.editors.manager.ui.fragments.main;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.DiffUtil;

import java.util.ArrayList;
import java.util.List;

import app.editors.manager.R;
import app.editors.manager.app.Api;
import app.editors.manager.mvp.models.account.Recent;
import app.editors.manager.mvp.models.base.Entity;
import app.editors.manager.mvp.models.explorer.Explorer;
import app.editors.manager.mvp.models.explorer.File;
import app.editors.manager.mvp.presenters.main.DocsBasePresenter;
import app.editors.manager.mvp.presenters.main.DocsRecentPresenter;
import app.editors.manager.mvp.views.main.DocsBaseView;
import app.editors.manager.mvp.views.main.DocsRecentView;
import app.editors.manager.ui.activities.base.BaseAppActivity;
import app.editors.manager.ui.activities.main.MainActivity;
import app.editors.manager.ui.activities.main.MediaActivity;
import app.editors.manager.ui.activities.main.WebViewerActivity;
import app.editors.manager.ui.adapters.RecentAdapter;
import app.editors.manager.ui.adapters.diffutilscallback.RecentDiffUtilsCallback;
import app.editors.manager.ui.dialogs.ContextBottomDialog;
import app.editors.manager.ui.views.custom.PlaceholderViews;
import lib.toolkit.base.managers.utils.StringUtils;
import moxy.presenter.InjectPresenter;

public class DocsRecentFragment extends DocsBaseFragment implements DocsRecentView, RecentAdapter.OnClick {

    public static String TAG = DocsRecentFragment.class.getSimpleName();

    public static DocsRecentFragment newInstance() {
        return new DocsRecentFragment();
    }

    private static final String KEY_FILTER = "KEY_FILTER";

    @InjectPresenter
    DocsRecentPresenter mDocsRecentPresenter;

    private MainActivity mMainActivity;
    private RecentAdapter mAdapter;
    private CharSequence mFilterValue;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof MainActivity) {
            mMainActivity = (MainActivity) context;
        } else {
            throw new RuntimeException(DocsRecentFragment.class.getSimpleName() + " - must implement - " +
                    MainActivity.class.getSimpleName());
        }
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (savedInstanceState != null && savedInstanceState.containsKey(KEY_FILTER)) {
            mFilterValue = savedInstanceState.getCharSequence(KEY_FILTER);
        }
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        if (mSearchView != null) {
            outState.putCharSequence(KEY_FILTER, mSearchView.getQuery());
        }
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return super.onCreateView(inflater, container, savedInstanceState);
    }

    @Override
    public boolean onBackPressed() {
        getActivity().invalidateOptionsMenu();
        return super.onBackPressed();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        init();
    }

    @Override
    public void onPrepareOptionsMenu(@NonNull Menu menu) {
        super.onPrepareOptionsMenu(menu);
        setMenuSearchEnabled(true);
        if (mMainItem != null) {
            mMainItem.setVisible(false);
        }
        if (mSortItem != null) {
            mSortItem.setVisible(true);
            mSortItem.setEnabled(true);
            mSortItem.getSubMenu().findItem(R.id.toolbar_sort_item_owner).setVisible(false);
//            mSortItem.getSubMenu().findItem(R.id.toolbar_sort_item_date_update).setChecked(true);
        }

    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        switch (requestCode) {
            case BaseAppActivity.REQUEST_ACTIVITY_WEB_VIEWER:
                mDocsRecentPresenter.getRecentFiles();
                break;
            case REQUEST_DOCS:
            case REQUEST_SHEETS:
            case REQUEST_PRESENTATION:
            case REQUEST_PDF:
                if (resultCode == Activity.RESULT_CANCELED) {
                    mDocsRecentPresenter.deleteTempFile();
                    break;
                } else if (resultCode == Activity.RESULT_OK) {
                    if (data.getData() != null) {
                        if (data.getBooleanExtra("EXTRA_IS_MODIFIED", false)) {
                            mDocsRecentPresenter.upload(data.getData(), null);
                        }
                    }
                }
                break;
        }
    }

    @Override
    public void onStateUpdateFilter(boolean isFilter, @Nullable String value) {
        super.onStateUpdateFilter(isFilter, value);
        if (isFilter) {
            mMainActivity.setAppBarStates(false);
            if (mSearchView != null) {
                mSearchView.setQuery(mFilterValue, true);
            }
        } else {
            mMainActivity.setAppBarStates(false);
            mMainActivity.setNavigationButton(true);
        }
    }

    @Override
    public void onRemoveItemFromFavorites() {

    }

    @Override
    protected Boolean isWebDav() {
        return false;
    }

    private void init() {
        mMainActivity.setAppBarStates(false);
        mMainActivity.setNavigationButton(true);
        mMainActivity.setActionButtonVisibility(false);
        mMainActivity.setOffToolbarAccount();

        mAdapter = new RecentAdapter(getContext());
        mAdapter.setOnClick(this);

        if (mRecyclerView != null) {
            mRecyclerView.setAdapter(mAdapter);
            mRecyclerView.setPadding(getResources().getDimensionPixelSize(R.dimen.screen_left_right_padding),
                    getResources().getDimensionPixelSize(R.dimen.screen_top_bottom_padding),
                    getResources().getDimensionPixelSize(R.dimen.screen_left_right_padding),
                    getResources().getDimensionPixelSize(R.dimen.screen_bottom_padding));
        }
        if (mSwipeRefresh != null) {
            mSwipeRefresh.setEnabled(false);
        }
        if (checkReadPermission()) {
            mDocsRecentPresenter.getRecentFiles();
        }

        setActionBarTitle(getString(R.string.fragment_recent_title));
    }

    @Override
    protected void onListEnd() {
        mDocsRecentPresenter.loadMore(mAdapter.getItemCount());
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (mMenu != null) {
            boolean isAscending = mMenu.findItem(R.id.toolbar_sort_item_asc).isChecked();
            switch (item.getItemId()) {
                case R.id.toolbar_item_sort:
                    mMainActivity.setAppBarStates(false);
                    mMainActivity.setNavigationButton(true);
                    break;
                case R.id.toolbar_sort_item_title:
                    if(item.isChecked()) {
                        mDocsRecentPresenter.reverseSortOrder(mAdapter.getItemList());
                    } else {
                        mDocsRecentPresenter.sortBy(Api.Parameters.VAL_SORT_BY_TITLE, isAscending);
                    }
                    item.setChecked(true);
                    break;
                case R.id.toolbar_sort_item_date_update:
                    if(item.isChecked()) {
                        mDocsRecentPresenter.reverseSortOrder(mAdapter.getItemList());
                    } else {
                        mDocsRecentPresenter.sortBy(Api.Parameters.VAL_SORT_BY_UPDATED, isAscending);
                    }
                    item.setChecked(true);
                    break;
                case R.id.toolbar_sort_item_owner:
                    if(item.isChecked()) {
                        mDocsRecentPresenter.reverseSortOrder(mAdapter.getItemList());
                    } else {
                        mDocsRecentPresenter.sortBy(Api.Parameters.VAL_SORT_BY_OWNER, isAscending);
                    }
                    item.setChecked(true);
                    break;
                case R.id.toolbar_sort_item_size:
                    if(item.isChecked()) {
                        mDocsRecentPresenter.reverseSortOrder(mAdapter.getItemList());
                    } else {
                        mDocsRecentPresenter.sortBy(Api.Parameters.VAL_SORT_BY_SIZE, isAscending);
                    }
                    item.setChecked(true);
                    break;
                case R.id.toolbar_sort_item_type:
                    if(item.isChecked()) {
                        mDocsRecentPresenter.reverseSortOrder(mAdapter.getItemList());
                    } else {
                        mDocsRecentPresenter.sortBy(Api.Parameters.VAL_SORT_BY_TYPE, isAscending);
                    }
                    item.setChecked(true);
                    break;
                case R.id.toolbar_sort_item_asc:
                case R.id.toolbar_sort_item_desc:
                    mDocsRecentPresenter.reverseList(mAdapter.getItemList(), isAscending);
                    break;
            }
        }
        item.setChecked(true);
        return false;
    }

    @Override
    public void onReverseSortOrder(List<Entity> itemList) {
        mAdapter.setData(itemList);
        mAdapter.notifyDataSetChanged();
        if(mMenu.findItem(R.id.toolbar_sort_item_desc).isChecked()) {
            mMenu.findItem(R.id.toolbar_sort_item_asc).setChecked(true);
        } else {
            mMenu.findItem(R.id.toolbar_sort_item_desc).setChecked(true);
        }
    }


    @Override
    public void updateFiles(List<Entity> files) {
        if (!files.isEmpty()) {
            if (mAdapter.getItemList() != null && mRecyclerView != null) {
                updateDiffUtils(files);
                mRecyclerView.scrollToPosition(0);
            } else {
                mAdapter.setItems(files);
            }
            mPlaceholderViews.setVisibility(false);
            updateMenu(true);
        } else {
            mPlaceholderViews.setTemplatePlaceholder(PlaceholderViews.Type.SEARCH);
            updateMenu(false);
        }
    }

    private void updateMenu(boolean isEnable) {
        if (mMenu != null && mSortItem != null && mSearchItem != null && mDeleteItem != null) {
            mSortItem.setEnabled(isEnable);
            mSearchItem.setEnabled(isEnable);
            mDeleteItem.setVisible(isEnable);
        }
    }

    private void updateDiffUtils(List<Entity> files) {
        RecentDiffUtilsCallback diffUtils = new RecentDiffUtilsCallback(files, mAdapter.getItemList());
        DiffUtil.DiffResult result = DiffUtil.calculateDiff(diffUtils);
        mAdapter.set(files, result);
    }

    @Override
    public void openFile(File file) {
        String ext = file.getFileExst();
        if (StringUtils.isVideoSupport(ext) || StringUtils.isImage(ext)) {
            MediaActivity.show(this, getExplorer(file), false);
        } else if (StringUtils.isDocument(ext)) {
            WebViewerActivity.show(requireActivity(), file);
        } else {
            onError(getString(R.string.error_unsupported_format));
        }
    }

    private Explorer getExplorer(File file) {
        Explorer explorer = new Explorer();
        List<File> files = new ArrayList<>();
        files.add(file);
        explorer.setFiles(files);
        return explorer;
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        if (requestCode == PERMISSION_READ_STORAGE) {
            if (grantResults.length == 1 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                mDocsRecentPresenter.getRecentFiles();
            } else {
                mPlaceholderViews.setTemplatePlaceholder(PlaceholderViews.Type.ACCESS);
            }
        }
    }

    @Override
    public boolean onQueryTextChange(String newText) {
        mSearchCloseButton.setVisibility(newText.isEmpty() ? View.INVISIBLE : View.VISIBLE);
        if (!newText.equals("")) {
            mDocsRecentPresenter.searchRecent(newText.toLowerCase());
        } else {
            mDocsRecentPresenter.getRecentFiles();
        }
        return false;
    }

    @Override
    public boolean onQueryTextSubmit(String query) {
        return false;
    }

    @Override
    public void onFileClick(Recent recent, int position) {
        mDocsRecentPresenter.fileClick(recent, position);
    }

    @Override
    public void onContextClick(Recent recent, int position) {
        mDocsRecentPresenter.contextClick(recent, position);
    }

    @Override
    public void onMoveElement(Recent recent, int position) {
        mAdapter.moveItem(position, 0);
        if (mRecyclerView != null) {
            mRecyclerView.scrollToPosition(0);
        }
    }

    @Override
    public void onContextShow(ContextBottomDialog.State state) {
        if (getFragmentManager() != null) {
            mContextBottomDialog.setState(state);
            mContextBottomDialog.show(getFragmentManager(), ContextBottomDialog.TAG);
        }
    }

    @Override
    public void onDeleteItem(int position) {
        mAdapter.removeItem(position);
    }

    @Override
    public void onContextButtonClick(ContextBottomDialog.Buttons buttons) {
        if (buttons == ContextBottomDialog.Buttons.DELETE) {
            mDocsRecentPresenter.deleteRecent();
        }
        mContextBottomDialog.dismiss();
    }

    @Override
    public void onOpenDocs(Uri uri) {
        showEditors(uri, EditorsType.DOCS);
    }

    @Override
    public void onOpenCells(Uri uri) {
        showEditors(uri, EditorsType.CELLS);
    }

    @Override
    public void onOpenPresentation(Uri uri) {
        showEditors(uri, EditorsType.PRESENTATION);
    }

    @Override
    public void onOpenPdf(Uri uri) {
        showEditors(uri, EditorsType.PDF);
    }

    @Override
    public void onOpenMedia(Explorer images, boolean isWebDav) {
        MediaActivity.show(this, images, isWebDav);
    }

    @Override
    protected DocsBasePresenter<? extends DocsBaseView> getPresenter() {
        return mDocsRecentPresenter;
    }
}

File: app_manager/appmanager/src/main/java/app/editors/manager/ui/fragments/main/DocsRecentFragment.kt
package app.editors.manager.ui.fragments.main

import android.Manifest
import android.app.Activity
import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.os.Environment
import android.provider.Settings
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import androidx.activity.result.contract.ActivityResultContracts
import androidx.annotation.RequiresApi
import app.documents.core.network.common.contracts.ApiContract
import app.documents.core.network.manager.models.explorer.CloudFile
import app.documents.core.network.manager.models.explorer.Explorer
import app.documents.core.storage.recent.Recent
import app.editors.manager.R
import app.editors.manager.app.appComponent
import app.editors.manager.mvp.presenters.main.DocsRecentPresenter
import app.editors.manager.mvp.presenters.main.OpenState
import app.editors.manager.mvp.presenters.main.RecentState
import app.editors.manager.mvp.views.main.DocsRecentView
import app.editors.manager.ui.activities.main.IMainActivity
import app.editors.manager.ui.activities.main.MainActivity
import app.editors.manager.ui.adapters.RecentAdapter
import app.editors.manager.ui.adapters.holders.factory.RecentHolderFactory
import app.editors.manager.ui.dialogs.explorer.ExplorerContextItem
import app.editors.manager.ui.dialogs.explorer.ExplorerContextState
import app.editors.manager.ui.popup.MainPopupItem
import app.editors.manager.ui.views.custom.PlaceholderViews
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import lib.toolkit.base.OpenMode
import lib.toolkit.base.managers.utils.LaunchActivityForResult
import lib.toolkit.base.managers.utils.RequestPermissions
import lib.toolkit.base.managers.utils.StringUtils
import lib.toolkit.base.managers.utils.TimeUtils
import lib.toolkit.base.managers.utils.UiUtils
import lib.toolkit.base.ui.dialogs.common.CommonDialog
import moxy.presenter.InjectPresenter
import java.util.Date

class DocsRecentFragment : DocsBaseFragment(), DocsRecentView {

    @InjectPresenter
    override lateinit var presenter: DocsRecentPresenter

    private var activity: IMainActivity? = null
    private var adapter: RecentAdapter? = null
    private var clearItem: MenuItem? = null

    private val recentListener: (recent: Recent) -> Unit = { recent ->
        Debounce.perform(1000L) { presenter.fileClick(recent, OpenMode.EDIT) }
    }

    private val contextListener: (recent: Recent, position: Int) -> Unit = { recent, position ->
        val state = ExplorerContextState(
            item = CloudFile().apply {
                title = recent.name
                fileExst = recent.name.split(".").let { if (it.size > 1) it[it.size - 1] else "" }
            },
            sectionType = ApiContract.Section.Recent.type,
            headerInfo = "${if (recent.isLocal) getString(R.string.this_device) else recent.source}" +
                    getString(R.string.placeholder_point) +
                    TimeUtils.formatDate(Date(recent.date))
        )
        presenter.onContextClick(recent, position)
        showExplorerContextBottomDialog(state)
    }

    private val readStorage = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) {
        presenter.recreateStack()
        presenter.getRecentFiles()
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        activity = if (context is IMainActivity) {
            context
        } else {
            throw RuntimeException(
                DocsRecentFragment::class.java.simpleName + " - must implement - " +
                        MainActivity::class.java.simpleName
            )
        }
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        if (isResumed) {
            clearItem = menu.findItem(R.id.toolbar_item_empty_trash)
            clearItem?.isVisible = true
            clearItem?.let { item ->
                UiUtils.setMenuItemTint(requireContext(), item, lib.toolkit.base.R.color.colorPrimary)
            }
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.toolbar_item_empty_trash -> {
                showClearDialog()
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        presenter.setSectionType(ApiContract.SectionType.CLOUD_RECENT)
        init()
    }

    override fun onPrepareOptionsMenu(menu: Menu) {
        super.onPrepareOptionsMenu(menu)
        setMenuSearchEnabled(true)
        mainItem?.isVisible = true
        clearItem?.let { item ->
            UiUtils.setMenuItemTint(requireContext(), item, lib.toolkit.base.R.color.colorPrimary)
        }
    }

    override fun onStateEmptyBackStack() {
        // stub
    }

    private fun init() {
        activity?.let { activity ->
            activity.setAppBarStates(false)
            activity.showNavigationButton(false)
            activity.showActionButton(false)
            activity.showAccount(false)
        }
        adapter = RecentAdapter(requireContext(), RecentHolderFactory(recentListener, contextListener))
        recyclerView?.let {
            it.adapter = adapter
            it.setPadding(
                resources.getDimensionPixelSize(lib.toolkit.base.R.dimen.screen_left_right_padding),
                resources.getDimensionPixelSize(lib.toolkit.base.R.dimen.screen_top_bottom_padding),
                resources.getDimensionPixelSize(lib.toolkit.base.R.dimen.screen_left_right_padding),
                resources.getDimensionPixelSize(lib.toolkit.base.R.dimen.screen_bottom_padding)
            )
        }
        swipeRefreshLayout?.isEnabled = false
        checkStorage()
        setActionBarTitle(getString(R.string.fragment_recent_title))
    }

    override fun onAcceptClick(dialogs: CommonDialog.Dialogs?, value: String?, tag: String?) {
        requestManage()
    }

    private fun checkStorage() {
        when {
            Build.VERSION.SDK_INT < Build.VERSION_CODES.R -> {
                requestReadWritePermission()
            }
            Build.VERSION.SDK_INT >= Build.VERSION_CODES.R -> {
                requestAccessStorage()
            }
        }
    }

    @RequiresApi(Build.VERSION_CODES.R)
    private fun requestAccessStorage() {
        if (!Environment.isExternalStorageManager() && requireContext().appComponent.preference.isShowStorageAccess) {
            showQuestionDialog(
                getString(R.string.app_manage_files_title),
                getString(R.string.app_manage_files_description),
                getString(R.string.dialogs_common_ok_button),
                getString(R.string.dialogs_common_cancel_button),
                TAG_STORAGE_ACCESS
            )
        } else {
            presenter.getRecentFiles()
        }
    }

    private fun requestReadWritePermission() {
        RequestPermissions(requireActivity().activityResultRegistry, { permissions ->
            if (permissions[Manifest.permission.WRITE_EXTERNAL_STORAGE] == true && permissions[Manifest.permission.READ_EXTERNAL_STORAGE] == true) {
                presenter.recreateStack()
                presenter.getRecentFiles()
            } else {
                swipeRefreshLayout?.isEnabled = false
                placeholderViews?.setTemplatePlaceholder(PlaceholderViews.Type.ACCESS)
            }
        }, arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE, Manifest.permission.READ_EXTERNAL_STORAGE)).request()
    }

    override fun updateFiles(files: List<Recent>, sortByUpdated: Boolean) {
        if (files.isNotEmpty()) {
            adapter?.setRecent(files, sortByUpdated)
            recyclerView?.scrollToPosition(0)
            placeholderViews?.setVisibility(false)
            updateMenu(true)
        } else {
            placeholderViews?.setTemplatePlaceholder(PlaceholderViews.Type.SEARCH)
            updateMenu(false)
        }
    }

    private fun updateMenu(isEnable: Boolean) {
        if (menu != null && searchItem != null && deleteItem != null) {
            searchItem?.isEnabled = isEnable
            deleteItem?.isVisible = isEnable
        }
    }

    override fun openFile(response: CloudFile, openMode: OpenMode) {
        val ext = response.fileExst
        if (StringUtils.isVideoSupport(ext) || StringUtils.isImage(ext)) {
            showMediaActivity(getExplorer(response), false) {
                // Stub
            }
        } else if (StringUtils.isDocument(ext)) {
            activity?.showWebViewer(response)
        } else {
            onError(getString(R.string.error_unsupported_format))
        }
    }

    private fun getExplorer(file: CloudFile): Explorer {
        return Explorer().apply {
            this.files = mutableListOf(file)
        }
    }

    override fun onDeleteItem(position: Int) {
        adapter?.let { recentAdapter ->
            recentAdapter.removeItem(position)
            if (recentAdapter.isEmpty()) setEmpty()
        }
    }

    override fun onContextButtonClick(contextItem: ExplorerContextItem) {
        when (contextItem) {
            is ExplorerContextItem.Edit -> presenter.fileClick(openMode = OpenMode.EDIT)
            is ExplorerContextItem.Preview -> presenter.fileClick(openMode = OpenMode.READ_ONLY)
            is ExplorerContextItem.Delete -> presenter.deleteRecent()
            else -> super.onContextButtonClick(contextItem)
        }
    }

    override fun onRender(state: RecentState) {
        when (state) {
            is RecentState.RenderList -> {
                if (state.recents.isEmpty()) {
                    setEmpty()
                } else {
                    setRecent(state.recents)
                }
            }
        }
    }

    private fun setRecent(recent: List<Recent>) {
        setMenuVisibility(true)
        presenter.updateFiles(recent)
    }

    private fun setEmpty() {
        setMenuVisibility(false)
        placeholderViews?.setTemplatePlaceholder(PlaceholderViews.Type.EMPTY)
    }

    override fun onStateUpdateFilter(isFilter: Boolean, value: String?) {
        super.onStateUpdateFilter(isFilter, value)
        activity?.showNavigationButton(isFilter)
    }

    override fun onOpenFile(state: OpenState) {
        hideDialog()
        when (state) {
            is OpenState.Docs, is OpenState.Cells, is OpenState.Slide, is OpenState.Pdf -> {
                LaunchActivityForResult(
                    requireActivity().activityResultRegistry,
                    { result ->
                        if (result.resultCode == Activity.RESULT_CANCELED) {
                            presenter.deleteTempFile()
                        } else if (result.resultCode == Activity.RESULT_OK) {
                            result.data?.data?.let {
                                if (result.data?.getBooleanExtra("EXTRA_IS_MODIFIED", false) == true) {
                                    presenter.upload(it, null)
                                }
                            }
                        }
                    },
                    getEditorsIntent(
                        uri = state.uri,
                        type = checkNotNull(state.type),
                        openMode = state.openMode
                    )
                ).show()
            }
            is OpenState.Media -> {
                showMediaActivity(state.explorer, state.isWebDav) {
                    // Stub
                }
            }
        }
    }

    override fun showMainActionPopup(vararg excluded: MainPopupItem) {
        super.showMainActionPopup(
            MainPopupItem.SortBy.Author,
            MainPopupItem.Select,
            MainPopupItem.SelectAll
        )
    }

    override val isWebDav: Boolean
        get() = false

    object Debounce {
        private var isClickable = true

        fun perform(timeMillis: Long, func: () -> Unit) {
            if (isClickable) {
                CoroutineScope(Dispatchers.Main).launch {
                    func.invoke()
                    isClickable = false
                    delay(timeMillis)
                    isClickable = true
                }
            }
        }
    }

    private fun showClearDialog() {
        UiUtils.showQuestionDialog(requireContext(), getString(R.string.clear_recents), acceptListener = {
            presenter.clearRecents()
        }, cancelListener = {
            // stub
        })
    }

    private fun requestManage() {
        try {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                readStorage.launch(
                    Intent(
                        Settings.ACTION_MANAGE_APP_ALL_FILES_ACCESS_PERMISSION,
                        Uri.parse("package:" + requireContext().packageName)
                    )
                )
            }
        } catch (e: ActivityNotFoundException) {
            openItem?.isVisible = false
            swipeRefreshLayout?.isEnabled = false
            activity?.showActionButton(false)
            placeholderViews?.setTemplatePlaceholder(PlaceholderViews.Type.ACCESS)
        }
    }

    companion object {
        var TAG: String = DocsRecentFragment::class.java.simpleName

        private const val TAG_STORAGE_ACCESS = "TAG_STORAGE_ACCESS"

        fun newInstance(): DocsRecentFragment {
            return DocsRecentFragment()
        }

    }

}